addons/account_facturx/models/account_invoice.py

```diff
@@ -0,0 +1,262 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, models, fields, tools, _
+from odoo.tools import DEFAULT_SERVER_DATE_FORMAT, pycompat, float_repr
+from odoo.tests.common import Form
+
+from datetime import datetime
+from lxml import etree
+from PyPDF2 import PdfFileReader
+from collections import namedtuple
+
+import io
+import base64
+
+
+DEFAULT_FACTURX_DATE_FORMAT = '%Y%m%d'
+
+
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+    _name = 'account.invoice'
+
+    @api.multi
+    def _export_as_facturx_xml(self):
+        ''' Create the Factur-x xml file content.
+        :return: The XML content as str.
+        '''
+        self.ensure_one()
+
+        def format_date(dt):
+            # Format the date in the Factur-x standard.
+            dt = dt or datetime.now()
+            return dt.strftime(DEFAULT_FACTURX_DATE_FORMAT)
+
+        def format_monetary(number, currency):
+            # Format the monetary values to avoid trailing decimals (e.g. 90.85000000000001).
+            return float_repr(number, currency.decimal_places)
+
+        # Create file content.
+        template_values = {
+            'record': self,
+            'format_date': format_date,
+            'format_monetary': format_monetary,
+        }
+        content = self.env.ref('account_facturx.account_invoice_facturx_export').render(template_values)
+        return b"<?xml version='1.0' encoding='UTF-8'?>" + content
+
+    @api.multi
+    def _import_facturx_invoice(self, tree):
+        ''' Extract invoice values from the Factur-x xml tree passed as parameter.
+
+        :param tree: The tree of the Factur-x xml file.
+        :return: A dictionary containing account.invoice values to create/update it.
+        '''
+        amount_total_import = None
+
+        # type must be present in the context to get the right behavior of the _default_journal method (account.invoice).
+        # journal_id must be present in the context to get the right behavior of the _default_account method (account.invoice.line).
+        self_ctx = self.with_context(type='in_invoice')
+        journal_id = self_ctx._default_journal().id
+        self_ctx = self_ctx.with_context(journal_id=journal_id)
+
+        # self could be a single record (editing) or be empty (new).
+        with Form(self_ctx, view='account.invoice_supplier_form') as invoice_form:
```

Comment:
@smetl @qdp-odoo @alexis-via @mart-e  I want to have your advice on that usage.

As the import is reflecting the usage (tests), is it a good way to do that in code which is not for ... tests ? Or is it a hack for the time being before v14 ?

Thanks for your answers!

Reply:
I think @xmo-odoo can answer as the main author of SSF, but I see this convenient for avoiding partial invoices. The import may be due to historical reasons.

Reply:
> The import may be due to historical reasons.

Which are ?

Reply:
It starts as a tool for tests, but then become solid thing and spread its usage, but didn't want to change all the existing imports (all of this is my lucubration).

Reply:
Yes, so, no formal usage, no docs for that, no use of real ORM to compute the onchanges, temporary usage. 

Moreover, you rely on code which is just disgusting to read : https://github.com/odoo/odoo/blob/13.0/odoo/tests/common.py#L1343

Only cons, ...

Reply:
Why do you say it's not documented? The classes have good docstrings. And there's real usage of ORM... I don't see cons.

Reply:
Yes, documented for a test usage...

Reply:
I don't think so, but let's give the voice to the Odoo people to pronounce. For now, 1 in favor and 1 against, hehe

Reply:
This usage is the easiest way to ensure a well formed invoice thanks to the onchanges. However, you are right: this code will be replaced by a simple `create` in the future, maybe for v14.0. We need to fix some remaining limitations of the current ORM before cleaning the onchanges.

Reply:
So, @pedrobaeza maybe https://github.com/OCA/server-tools/tree/13.0/onchange_helper is cleaner way to do what we want as it pull less code... see https://github.com/OCA/server-tools/blob/13.0/onchange_helper/models/base.py#L8

Reply:
@rousseldenis we all agree that this is due to the current limitations on the ORM and the duality onchanges/compute, so it's NEEDED. Now, how to solve it?

1. Using an extra module with all its maintenance burden (onchange_helper)?
2. Using an already tested feature on Odoo itself that allows to collaborate in the maintenance (odoo.tests.Form) and the tested base is bigger?

I clearly prefer 2. If you find any problem with SSF, let's focus on solve it (in current state, I have to say it's pretty good, though there's only one little rough edge that I will notify soon with related fields over the inverse field in one2many).

Reply:
About the cleaner way, I disagree, as with `onchange_helper` you need to know in advance which fields you want to return, so that disables the inheritance compatibility with third party modules that add fields and react on the same onchanges.

Reply:
> About the cleaner way, I disagree, as with `onchange_helper` you need to know in advance which fields you want to return, so that disables the inheritance compatibility with third party modules that add fields and react on the same onchanges.

That's a misunderstanding of the module, as you don't need to know the fields...

Reply:
Good to know, but even that, I prefer to not add extra dependencies and maintenance burden for getting the same result.